File: string/leetcode_345.java
package string;

import java.util.HashSet;
import java.util.Stack;

public class leetcode_345 {
    public static void main(String[] args) {
        leetcode_345 lc = new leetcode_345();
        System.out.println(lc.reverseVowels("hello"));
    }

    static HashSet<Character> vowels = new HashSet<>();

    static {
        vowels.add('a');
        vowels.add('A');
        vowels.add('e');
        vowels.add('E');
        vowels.add('i');
        vowels.add('I');
        vowels.add('o');
        vowels.add('O');
        vowels.add('u');
        vowels.add('U');
    }

    public String reverseVowels(String s) {
        if (s.length() == 0) return s;

        Stack<Character> stack = new Stack<>();
        StringBuilder sb = new StringBuilder();

        for (char c : s.toCharArray()) {
            if (vowels.contains(c)) stack.push(c);
        }

        for (char c : s.toCharArray()) {
            if (vowels.contains(c))
                sb.append(stack.pop());
            else
                sb.append(c);
        }

        return sb.toString();
    }
}

File: string/LeetCode_345.kt
package string

class LeetCode_345 {

    /**
     * Solution : Two Pointers
     * Time Complexity : O(N)
     * Space Complexity : O(N)
     *
     * @param s string
     * @return return string which reverse only all the vowels in it
    */
    fun reverseVowels(s: String): String {
        val vowelSet = setOf('a', 'e', 'o', 'u', 'i', 'A', 'E', 'O', 'U', 'I')

        var i = 0
        var j = s.lastIndex

        val charArray = s.toCharArray()

        while (i < j) {
            val firstChar = s[i]
            val secondChar = s[j]

            if (vowelSet.contains(firstChar) && vowelSet.contains(secondChar)) {
                swap(charArray, i, j)
                i++
                j--
            } else if (vowelSet.contains(firstChar)) {
                j--
            } else if (vowelSet.contains(secondChar)) {
                i++
            } else {
                i++
                j--
            }
        }

        return charArray.concatToString()
    }

    private fun swap(charArray: CharArray, x: Int, y: Int) {
        val tmp = charArray[x]
        charArray[x] = charArray[y]
        charArray[y] = tmp
    }
}

fun main() {
    // Input
    val s = "hello"
    // Output
    val result = LeetCode_345().reverseVowels(s)
    println(result)
    require(result == "holle")
}
